contracts/src/v0.8/shared/call/CallWithExactGas.sol

```diff
@@ -0,0 +1,161 @@
+    bytes memory payload,
+    address target,
+    uint256 gasLimit,
+    uint16 gasForCallExactCheck,
+    uint16 maxReturnBytes
+  ) internal returns (bool success, bytes memory retData, uint256 gasUsed) {
+    // allocate retData memory ahead of time
+    retData = new bytes(maxReturnBytes);
+
+    assembly {
+      // solidity calls check that a contract actually exists at the destination, so we do the same
+      // Note we do this check prior to measuring gas so gasForCallExactCheck (our "cushion")
+      // doesn't need to account for it.
+      if iszero(extcodesize(target)) {
+        mstore(0, NO_CONTRACT_SIG)
+        revert(0, 0x4)
+      }
+
+      let g := gas()
+      // Compute g -= gasForCallExactCheck and check for underflow
+      // The gas actually passed to the callee is _min(gasAmount, 63//64*gas available).
+      // We want to ensure that we revert if gasAmount >  63//64*gas available
+      // as we do not want to provide them with less, however that check itself costs
+      // gas. gasForCallExactCheck ensures we have at least enough gas to be able
+      // to revert if gasAmount >  63//64*gas available.
+      if lt(g, gasForCallExactCheck) {
+        mstore(0, NO_GAS_FOR_CALL_EXACT_CHECK_SIG)
+        revert(0, 0x4)
+      }
+      g := sub(g, gasForCallExactCheck)
+      // if g - g//64 <= gasAmount, revert. We subtract g//64 because of EIP-150
+      if iszero(gt(sub(g, div(g, 64)), gasLimit)) {
+        mstore(0, NOT_ENOUGH_GAS_FOR_CALL_SIG)
+        revert(0, 0x4)
+      }
+
+      // We save the gas before the call so we can calculate how much gas the call used
+      let gasBeforeCall := gas()
+      // call and return whether we succeeded. ignore return data
+      // call(gas,addr,value,argsOffset,argsLength,retOffset,retLength)
+      success := call(gasLimit, target, 0, add(payload, 0x20), mload(payload), 0, 0)
+      gasUsed := sub(gasBeforeCall, gas())
+
+      // limit our copy to maxReturnBytes bytes
+      let toCopy := returndatasize()
+      if gt(toCopy, maxReturnBytes) {
+        toCopy := maxReturnBytes
+      }
+      // Store the length of the copied bytes
+      mstore(retData, toCopy)
+      // copy the bytes from retData[0:_toCopy]
+      returndatacopy(add(retData, 0x20), 0, toCopy)
+    }
+    return (success, retData, gasUsed);
+  }
+
+  /// @notice Calls target address with exactly gasAmount gas and payload as calldata
+  /// or reverts if at least gasLimit gas is not available.
+  /// @dev Does not check if target is a contract. If it is not a contract, the low-level
+  /// call will still be made and it will succeed.
+  /// @dev Ignores the return data, which makes it immune to gas bomb attacks.
+  /// @return sufficientGas Whether there was enough gas to make the call
+  function _callWithExactGasEvenIfTargetIsNoContract(
+    bytes memory payload,
+    address target,
+    uint256 gasLimit,
+    uint16 gasForCallExactCheck
+  ) internal returns (bool sufficientGas) {
```

Comment:
I wonder whether the code be made more misuse resistant here. All other library functions return `success`, not `sufficientGas`. One could easily mistake it and it would be hard for a reviewer to catch.